### PR TITLE
Profile credentials dialer fix

### PR DESCRIPTION
**Pull request by Joerger**
Change Profile dialer to make insecure connection when retrieving web proxy address.

### CODE REVIEW

**Comment by Joerger** [comment]
@awly @andrejtokarcik Please take a look at this one line bug fix ASAP, thanks.

**Review by awly — APPROVED** [review]
In the future we should store this as a field in the user profile.
If the user didn't pass `--insecure` during `tsh login`, we should generally verify the certs

**Review by klizhentas — CHANGES_REQUESTED** [review]
Before we merge this, so I understand this sets the first request to use insecure connection. What is the rationale for this?

**Comment by Joerger** [comment]
> Before we merge this, so I understand this sets the first request to use insecure connection. What is the rationale for this?

The insecure connection is being made to the web client just to retrieve the tunnel proxy address. I modeled this logic after `tctl`, which also initiates an insecure connection to the web client to retrieve the tunnel proxy address.

**Comment by klizhentas** [comment]
@Joerger thanks. I understand that both tctl and now client are making first insecure connection to get a proxy address from a proxy. Any reason we can't fix it now?

**Comment by Joerger** [comment]
I would need to look into it more and update the tsh profile to hold TLS for the web client. I tried using the TLS certificates already generated in the profile, but it seems that the web client has a different certificate authority. I could be wrong and it will be a simple fix (especially for someone who knows their way around the tsh code), but I suspect it would set us back on our 6.1 release, especially since I'm still trying up some other work for the API. I am definitely willing to try, but a patch release seems to have fewer downsides, unless this is a heavier security risk than I am understanding.

**Comment by awly** [comment]
IIRC, this is how proxy address discovery worked before the client refactor - we always hit ping/find endpoints with `InsecureSkipVerify`.
The proxy uses a self-signed cert on the HTTP API if none was supplied; we don't have a CA for it available as with host SSH and TLS certs.

Let's make `InsecureAddressDiscovery` a field of `ProfileCreds` for now, so the caller has to opt-in.

**Review by klizhentas — APPROVED** [review]

**Comment by klizhentas** [comment]
@awly and @Joerger thanks team for digging into this and explaining it to me

File: api/client/client.go [hunk 1]
```diff
@@ -337,6 +337,10 @@ type Config struct {
 	// rather than blocking until the connection is up. A predefined Dialer
 	// or an auth server address must be provided.
 	DialInBackground bool
+	// InsecureAddressDiscovery specifies that the client can try to retrieve
+	// a tunnel proxy address from the web proxy using an insecure connection.
+	// Currently, the web proxy can not be used without an insecure connection.
```
Comment (awly) [inline api/client/client.go:342]:
```suggestion
	// The web proxy uses a self-signed TLS certificate by default, which requires this field to be set.
	// If the web proxy was provided with signed TLS certificates, this field should not be set.
```

File: api/client/credentials.go [hunk 2]
```diff
@@ -201,16 +200,20 @@ type ProfileCreds struct {
 }
 
 // Dialer is used to dial a connection to Auth.
-func (c *ProfileCreds) Dialer(keepAliveInterval, dialTimeout time.Duration) (ContextDialer, error) {
+func (c *ProfileCreds) Dialer(cfg Config) (ContextDialer, error) {
+	if !cfg.InsecureAddressDiscovery {
+		return nil, trace.NotImplemented("cannot retrieve tunnel proxy address without an insecure connection")
+	}
```
Comment (awly) [inline api/client/credentials.go:206]: This is not required if the user provides the proxy with signed TLS certs

File: api/client/credentials.go [hunk 3]
```diff
@@ -201,16 +200,20 @@ type ProfileCreds struct {
...
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	dialer := NewTunnelDialer(*sshConfig, keepAliveInterval, dialTimeout)
+	dialer := NewTunnelDialer(*sshConfig, cfg.KeepAlivePeriod, cfg.DialTimeout)
 	return ContextDialerFunc(func(ctx context.Context, network, _ string) (conn net.Conn, err error) {
 		// Ping web proxy to retrieve tunnel proxy address.
-		pr, err := Find(ctx, c.profile.WebProxyAddr, false, nil)
+		pr, err := Find(ctx, c.profile.WebProxyAddr, true, nil)
```
Comment (awly) [inline api/client/credentials.go:216]:
```suggestion
		pr, err := Find(ctx, c.profile.WebProxyAddr, cfg.InsecureAddressDiscovery, nil)
```

**Review by awly — COMMENTED** [review]
Could you also rebase and re-run `make update-vendor`, since there seem to be some `go.mod` issues

File: api/client/credentials.go (hunk 3 above)
Comment (Joerger) [reply api/client/credentials.go:216]: If insecure is false, doesn't a valid `*x509.CertPool` need to be provided in place of nil?

Comment (awly) [reply api/client/credentials.go:216]: If `CertPool` is nil, `crypto/tls` will use the system pool

**Pull request merged**